File: withargs/which.py
import fnmatch
import os
import sys

def usage() :
    print 'which -p <command>'
    exit()


def check_exact(dir, file2find) :
    if show_dirs :
        print 'checking in ' + dir
    try:
        for filename in os.listdir(dir):
            if filename == file2find:
                return True
        return False
    except Exception as e:
        print e
        return False


def find_in_dirs(dirs, file2find) :
    for dir in dirs :
        if dir == '':
            continue
        result = check_exact(dir, file2find)
        if result == True :
            print 'found in ' + dir + '\\' + file2find
            return True
    print file2find + ' not found.'
    return False



dirs  = (os.environ['PATH']).split(';')
show_dirs = False
file2find = ''

if len(sys.argv) == 3:
    if sys.argv[1] == '-p':
        show_dirs = True
        file2find = sys.argv[2]
    else :
        usage()
else :
    file2find = sys.argv[1]
    

find_in_dirs(dirs, file2find)

File: withargs/which.php
<?php

///////////// main ///////////////

$cmd = null;
$show_dirs = false;

get_params($argv);
$cleaned_dirs = get_path_dirs();
$result = search_in_dirs($cleaned_dirs, $cmd, $show_dirs);
if ($result != null)
    echo "found at $result\n";
else
    echo "command $cmd not found.\n";
exit (($result != null) ? 0 : 1);

////////////////////////

/* search for $cmd in $dirs, optionally show directories searched */
function search_in_dirs($dirs, $cmd, $show_dirs)
{
    foreach($dirs as $dir){
        if ($show_dirs)
            echo "$dir\n";

        if (check_exact($dir, $cmd))
            return $dir;
    }

    return null;
}

/* look for an exact filename match on the files in $dir */
function check_exact($dir, $cmd)
{
    $result = false;
    if ($handle = @opendir($dir)) {
        while (false !== ($entry = readdir($handle))) {
            if ($entry == $cmd){
                $result = true;
                break;
            }
        }
        closedir($handle);
    }else {
        echo "unable to open $dir\n";
    }

    return $result;
}

/* get an array of directory paths from the path environment variable string */
function get_path_dirs ()
{
    $path_string = getenv("PATH");
    $path_dirs = explode(';', $path_string);
    $cleaned_dirs = array_filter( $path_dirs, function ($dir){
        return ($dir != null && $dir != "");
    });

    return $cleaned_dirs;
}

/* learn 'em */
function usage() {
    echo "usage: php which.php [-p] <command>\n";
    exit(1);
}

/* parse command line arguments */
function get_params($argv){
    global $cmd;
    global $show_dirs;

    if (sizeof($argv) < 2)
        usage();

    for ($i = 1; $i < sizeof($argv); $i++){
        $arg = $argv[$i];
        if ($arg == '-p')
            $show_dirs = true;
        else if ($cmd == null)
            $cmd = $arg;
        else{
            echo "invalid argument: $arg\n";
            usage();
        }
    }

    if ($cmd == null)
        usage();
}


?>